src/dune_rules/warning_emit.mli

```diff
@@ -0,0 +1,13 @@
+(** Emit warnings that respect dune's conventions *)
+
+open Import
+
+type context =
+  | Build of Path.Build.t
+  | Source of Source_tree.Dir.t
+  | Source_dir of Path.Source.t
+  | Dune_project of Dune_project.t
+
+(** [emit w ctx f] will call [f] if [w] is enabled in [ctx]. [f] should
+    generate the warning corresponding to [w] *)
+val emit : Warning.t -> context -> (unit -> unit Memo.t) -> unit Memo.t
```

Comment:
If the function `f` is expected to always print a warning message then why not have the `emit` function take warning message directly and print it itself? To avoid computing the message in the common case when it's not needed it could take a thunk which returns the message.